File: Code/dotNetBasics/ConsoleApp/Person/PersonFluent.cs
using System.Net.NetworkInformation;

namespace ConsoleApp.Person
{
    enum PersonStatus
    {
        UNKNOWN = 0, NEW, INITING, READY
    }

    /// <summary>
    /// This class uses fluent initialization.
    /// </summary>
    internal class PersonFluent
    {
        private string _name;

        private bool _immutable;
        private PersonCfg _cfg;

        private PersonStatus _status = PersonStatus.UNKNOWN;

        public PersonFluent()
        {
            _cfg = new PersonCfg();
            _status = PersonStatus.NEW;
            _name = string.Empty;
        }

        /// <summary>
        /// To hide away the two different implementations
        /// </summary>
        /// <returns></returns>
        private bool isImmutable() => _cfg.Immutable;

        /// <summary>
        /// Clone this, if we are immutable.
        /// Note: there are better ways to do this..
        /// </summary>
        /// <returns></returns>
        private PersonFluent CloneThis()
        {
            PersonFluent p = new PersonFluent();
            p._name = this._name;
            p._status = this._status;
            p._cfg = this._cfg;
            // ...other fields
            return p;
        }

        /// <summary>
        /// Check for the consistency
        /// </summary>
        /// <param name="wanted"></param>
        private void ThrowIfStatusIsNot(PersonStatus wanted)
        {
            if (_status != wanted)
                throw new Exception($"Operation permitted only in status {wanted}. Found {_status}");
        }

        /// <summary>
        /// Version with params. Each time you have a new configuration option you need to add a param to this fn.
        /// This must be the first method that you call on your chain.
        /// </summary>
        /// <param name="immutable"></param>
        /// <returns></returns>
        public PersonFluent Config(bool immutable)
        {
            ThrowIfStatusIsNot(PersonStatus.NEW);
		
		    _immutable = immutable;
		
		    _status = PersonStatus.INITING;
		    return isImmutable() ? CloneThis() : this;
        }

        /// <summary>
        /// Version with Cfg obj. This is more scalable.
        /// This must be the first method that you call on your chain.
        /// </summary>
        /// <param name="lambda"></param>
        /// <returns></returns>
        public PersonFluent Config(Action<PersonCfg> lambda)
        {
            ThrowIfStatusIsNot(PersonStatus.NEW);
		
		    // Let the lambda visit my config object
		    lambda(_cfg);
		
		    _status = PersonStatus.INITING;
		
		    return isImmutable()? CloneThis() : this;
        }

        /// <summary>
        /// Set Name (fluent)
        /// </summary>
        /// <param name="name"></param>
        /// <returns></returns>
        public PersonFluent WithName(string name)
        {
            ThrowIfStatusIsNot(PersonStatus.INITING);
		
		    _name = name;
		
		    _status = PersonStatus.INITING;
		
		    return isImmutable() ? CloneThis() : this;
        }

        /// <summary>
        /// This must be the last method that you call on your chain.
        /// </summary>
        /// <returns></returns>
        public PersonFluent Persist()
        {
            ThrowIfStatusIsNot(PersonStatus.INITING);
		
		    // Do something.

		    _status = PersonStatus.READY;
		
		    return isImmutable() ? CloneThis() : this;
        }

        public override string ToString() => $"Name: {_name}";
    }
}

File: Code/Fluent.Java/src/person/PersonFluent.java
package person;

/**
 * This class uses fluent initialization
 */
public class PersonFluent {
	
	private String _name = null;
	
	private Boolean _immutable = false;
	private PersonCfg _cfg = null;
	
	private PersonStatus _status = PersonStatus.UNKNOWN;
	
	public PersonFluent() {
		this._cfg = new PersonCfg();
		this._status = PersonStatus.NEW;
	}
	
	/**
	 * To hide away the two different implementations
	 * @return
	 */
	private Boolean isImmutable() {
		return this._cfg.Immutable;
	}
	
	/**
	 * Clone this, if we are immutable.
	 * Note: there are better ways to do this..
	 * @return
	 */
	private PersonFluent cloneThis() {
		PersonFluent p = new PersonFluent();
		p._name = this._name;
		p._status = this._status;
		p._cfg = this._cfg;
		// ...other fields
		return p;
	}
	
	/*
	 * Check for the consistency
	 */
	private void throwIfStatusIsNot(PersonStatus wanted) throws Exception {
		if(this._status != wanted)
			throw new Exception("Operation permitted only in status "
				+ wanted + ". Found " + this._status);
	}
	
	/**
	 * Version with params. Each time you have a new configuration option
	 * you need to add a param to this fn.
	 * This must be the first method that you call on your chain.
	 * @param immutable
	 * @return
	 * @throws Exception
	 */
	public PersonFluent Config(Boolean immutable) throws Exception {
		throwIfStatusIsNot(PersonStatus.NEW);
		
		this._immutable = immutable;
		
		this._status = PersonStatus.INITING;
		
		
		return isImmutable() ? cloneThis() : this;
	}
	
	/**
	 * Version with Cfg obj. This is more scalable.
	 * This must be the first method that you call on your chain.
	 * @param lambda
	 * @return
	 * @throws Exception
	 */
	public PersonFluent Config(ILambdaWithTemplate<PersonCfg> lambda) throws Exception {
		throwIfStatusIsNot(PersonStatus.NEW);
		
		// Let the lambda visit my config object
		lambda.run(this._cfg);
		
		this._status = PersonStatus.INITING;
		
		return isImmutable() ? cloneThis() : this;
	}
	
	/**
	 * Set Name (fluent)
	 */
	public PersonFluent withName(String name) throws Exception {
		throwIfStatusIsNot(PersonStatus.INITING);
		
		this._name = name;
		
		this._status = PersonStatus.INITING;
		
		return isImmutable() ? cloneThis() : this;
	}
	
	/**
	 * This must be the last method that you call on your chain.
	 * @return
	 * @throws Exception
	 */
	public PersonFluent Persist() throws Exception {
		throwIfStatusIsNot(PersonStatus.INITING);
		
		// Do something.

		this._status = PersonStatus.READY;
		
		return isImmutable() ? cloneThis() : this;
	}
	
	@Override
	public String toString() {
		
		return "Name: " + this._name;
	}

}
